presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/JoinGraphNode.java

```diff
@@ -64,22 +72,67 @@ public JoinGraphNode(PlanNodeId id, List<PlanNode> nodes, List<JoinNode.EquiJoin
     @Override
     public List<PlanNode> getSources()
     {
-        return ImmutableList.copyOf(nodes);
+        return sources;
     }
 
     @Override
     public List<Symbol> getOutputSymbols()
     {
-        ImmutableList.Builder<Symbol> symbolBuilder = ImmutableList.builder();
-        for (PlanNode source : nodes) {
-            symbolBuilder.addAll(source.getOutputSymbols());
-        }
-        return symbolBuilder.build();
+        return sources.stream()
+                .flatMap(source -> source.getOutputSymbols().stream())
+                .collect(toImmutableList());
     }
 
     @Override
     public PlanNode replaceChildren(List<PlanNode> newChildren)
     {
-        return new JoinGraphNode(getId(), newChildren, criteria, filters);
+        return new JoinGraphNode(getId(), newChildren, filter);
+    }
+
+    public String getKey()
```

Comment:
I think string key could hurt performance if we want to enumerate though many join reorderings. You could simply implement `equal` and `hashCode`? I know that plan nodes don't provide these, but identity comparison of `sources` should be fine. Still filters would need to be compared by identity. 

Add `TODO` to provide better key implementation